src/component/common/ConstraintAccordion/ConstraintAccordionList/ConstraintAccordionList.tsx

```diff
@@ -0,0 +1,142 @@
+import { IConstraint } from 'interfaces/strategy';
+import React, { forwardRef, useImperativeHandle } from 'react';
+import { ConstraintAccordion } from 'component/common/ConstraintAccordion/ConstraintAccordion';
+import produce from 'immer';
+import useUnleashContext from 'hooks/api/getters/useUnleashContext/useUnleashContext';
+import PermissionButton from 'component/common/PermissionButton/PermissionButton';
+import {
+    CREATE_FEATURE_STRATEGY,
+    UPDATE_FEATURE_STRATEGY,
+} from 'component/providers/AccessProvider/permissions';
+import { useWeakMap } from 'hooks/useWeakMap';
+import { objectId } from 'utils/object-id';
+import { useStyles } from './ConstraintAccordionList.styles';
+import { createEmptyConstraint } from 'component/common/ConstraintAccordion/ConstraintAccordionList/createEmptyConstraint';
+import ConditionallyRender from 'component/common/ConditionallyRender';
+
+interface IConstraintAccordionListProps {
+    projectId?: string;
+    environmentId?: string;
+    constraints: IConstraint[];
+    setConstraints: React.Dispatch<React.SetStateAction<IConstraint[]>>;
+    showCreateButton?: boolean;
+}
+
+// Ref methods exposed by this component.
+export interface IConstraintAccordionListRef {
+    addConstraint: (contextName: string) => void;
+}
+
+// Extra form state for each constraint.
+interface IConstraintAccordionListItemState {
+    // Is the constraint new (never been saved)?
+    new?: boolean;
+    // Is the constraint currently being edited?
+    editing?: boolean;
+}
+
+export const ConstraintAccordionList = forwardRef<
+    IConstraintAccordionListRef | undefined,
+    IConstraintAccordionListProps
+>(
+    (
+        {
+            projectId,
+            environmentId,
+            constraints,
+            setConstraints,
+            showCreateButton,
+        },
+        ref
+    ) => {
+        const state = useWeakMap<
+            IConstraint,
+            IConstraintAccordionListItemState
+        >();
+        const { context } = useUnleashContext();
+        const styles = useStyles();
+
+        const addConstraint = (contextName: string) => {
+            const constraint = createEmptyConstraint(contextName);
+            state.set(constraint, { editing: true, new: true });
+            setConstraints(prev => [...prev, constraint]);
+        };
+
+        useImperativeHandle(ref, () => ({
+            addConstraint,
+        }));
+
+        const onAdd = () => {
+            addConstraint(context[0].name);
+        };
+
+        const onEdit = (constraint: IConstraint) => {
+            state.set(constraint, { editing: true });
+        };
+
+        const onCancel = (index: number) => {
+            const constraint = constraints[index];
+            state.get(constraint)?.new && onRemove(index);
+            state.set(constraint, {});
+        };
+
+        const onRemove = (index: number) => {
+            const constraint = constraints[index];
+            state.set(constraint, {});
```

Comment:
If I'm understanding this correctly, this WeakMap is a key value store where the object is used as the key and the metadata is associated with the object reference? Does it trigger re-renders?

I'm curious about this pattern.

Reply:
Yeah, just a way to associate extra fields with an object, without extending the object itself. The `useWeakMap` hook takes care of triggering re-renders whenever the map updates.